### PR TITLE
replace ark's map_utils.go with structured types and apimachinery's unstructured helpers

### PR DESCRIPTION
**Pull request by skriss**
Long-awaited update to #724, which I'm closing. 

The goals of this PR are to (a) eliminate our `pkg/util/collections/map_utils.go` since the functionality is largely now available in upstream code; and (b) where possible, simplify our code by converting unstructured representations to typed objects for querying/manipulating data.

Units are all passing which makes me feel pretty good, but am planning to do some manual testing as well.

### CODE REVIEW

File: pkg/restore/pod_action.go [hunk 1]
```diff
@@ -41,94 +43,213 @@ func (a *podAction) AppliesTo() (ResourceSelector, error) {
...
-func removeVolumeMounts(podSpec map[string]interface{}, containersKey, prefix string, log logrus.FieldLogger) error {
-	return collections.ForEach(podSpec, containersKey, func(container map[string]interface{}) error {
-		if !collections.Exists(container, "volumeMounts") {
-			return nil
-		}
-
-		return removeItemsWithNamePrefix(container, "volumeMounts", prefix, log)
-	})
-}
+// func (a *podAction) Execute(obj runtime.Unstructured, restore *api.Restore) (runtime.Unstructured, error, error) {
```
Comment (nrb) [inline pkg/restore/pod_action.go]: Should this commented section be removed?

File: pkg/restore/service_action.go [hunk 2]
```diff
@@ -82,31 +88,65 @@ func getPreservedPorts(obj runtime.Unstructured) (map[string]bool, error) {
...
+	for i, port := range service.Spec.Ports {
+		if !explicitNodePorts.Has(port.Name) {
+			service.Spec.Ports[i].NodePort = 0
 		}
-		delete(p, "nodePort")
 	}
+
 	return nil
+
+	// preservedPorts, err := getPreservedPorts(obj)
```
Comment (nrb) [inline pkg/restore/service_action.go]: Should this commented section be deleted?

File: pkg/restore/service_action.go [hunk 3]
```diff
@@ -46,20 +47,25 @@ func (a *serviceAction) AppliesTo() (ResourceSelector, error) {
...
+
+	res, err := runtime.DefaultUnstructuredConverter.ToUnstructured(service)
+	if err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+
+	return &unstructured.Unstructured{Object: res}, nil, nil
 }
 
 func getPreservedPorts(obj runtime.Unstructured) (map[string]bool, error) {
```
Comment (ncdc) [inline pkg/restore/service_action.go]: Do we need this func any more?

File: pkg/restore/pod_action.go [hunk 4]
```diff
@@ -41,94 +43,213 @@ func (a *podAction) AppliesTo() (ResourceSelector, error) {
...
-	if err := removeItemsWithNamePrefix(spec, "volumes", prefix, a.logger); err != nil {
-		return nil, nil, err
-	}
-
-	// remove the service account token volume mount from all containers
-	a.logger.Debug("iterating over containers")
-	if err := removeVolumeMounts(spec, "containers", prefix, a.logger); err != nil {
-		return nil, nil, err
+	if len(pod.Spec.Volumes) == 0 {
+		res, err := runtime.DefaultUnstructuredConverter.ToUnstructured(pod)
```
[the remark this answers is not in the document]
Comment (nrb) [reply pkg/restore/pod_action.go]: I like this suggestion

[the remark this answers is not in the document]
Comment (skriss) [reply pkg/restore/pod_action.go]: eh, I guess all pods are going to have a service account token volume, so the `if` is kind of a pointless optimization anyway. i'll just remove it and always go through the loops. We're not talking about a large number of iterations anyway, so no performance concerns.

File: pkg/restore/restore.go [hunk 5]
```diff
@@ -1011,9 +1011,13 @@ func (r *pvRestorer) executePVAction(obj *unstructured.Unstructured) (*unstructu
 		return nil, errors.New("PersistentVolume is missing its name")
 	}
 
-	spec, err := collections.GetMap(obj.UnstructuredContent(), "spec")
-	if err != nil {
-		return nil, errors.WithStack(err)
+	res, ok := obj.Object["spec"]
```
[the remark this answers is not in the document]
Comment (skriss) [reply pkg/restore/restore.go]: yeah. adding a comment.

File: pkg/restore/service_action.go (hunk 3 above)
Comment (skriss) [reply pkg/restore/service_action.go]: nope

File: pkg/controller/restore_controller_test.go [hunk 6]
```diff
@@ -439,11 +439,15 @@ func TestProcessRestore(t *testing.T) {
 						return false, nil, err
 					}
 
-					phase, err := collections.GetString(patchMap, "status.phase")
+					phase, found, err := unstructured.NestedString(patchMap, "status", "phase")
 					if err != nil {
 						t.Logf("error getting status.phase: %s\n", err)
 						return false, nil, err
```
[the remark this answers is not in the document]
Comment (skriss) [reply pkg/controller/restore_controller_test.go]: eh, it's test code, don't think it really matters?

**Comment by skriss** [comment]
addressed code review comments

File: pkg/controller/restore_controller_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/controller/restore_controller_test.go]: Missed that fact; nevermind

**Comment by skriss** [comment]
rebased, but don't want this to go in until v0.12

**Comment by skriss** [comment]
@nrb @carlisia when you get a chance, this is ready for another review, and merge when ready.

**Comment by skriss** [comment]
I probably owe a followup changelog on this one given the wide impact

**Pull request merged**